src/main/java/com/meta/chatbridge/message/WAMessageHandler.java

```diff
@@ -0,0 +1,222 @@
+        }
+        case POST -> {
+          return postHandler(ctx);
+        }
+      }
+    } catch (RuntimeException e) {
+      LOGGER.error(e.getMessage(), e);
+      throw e;
+    } catch (Exception e) {
+      LOGGER.error(e.getMessage(), e);
+      throw new RuntimeException(e);
+    }
+    throw new UnsupportedOperationException("Only accepting get and post methods");
+  }
+
+  List<WAMessage> postHandler(Context ctx) {
+    MetaHandlerUtils.postHeaderValidator(ctx, appSecret);
+    String bodyString = ctx.body();
+    WebhookPayload payload;
+    try {
+      payload = MAPPER.readValue(bodyString, WebhookPayload.class);
+    } catch (Exception e) {
+      LOGGER
+          .atWarn()
+          .addKeyValue("body", bodyString)
+          .setMessage(
+              "unable to process message, server may be subscribed to a 'webhook field' it cannot process")
+          .log();
+      throw new RuntimeException(e);
+    }
+
+    List<WAMessage> waMessages = new ArrayList<>();
+    payload.entry().stream()
+        .flatMap(e -> e.changes().stream())
+        .forEach(
+            change -> {
+              Identifier phoneNumberId = change.value().metadata().phoneNumberId();
+              for (WebhookMessage message : change.value().messages()) {
+                if (messageDeduplicator.addAndGetIsDuplicate(message.id())) {
+                  continue; // message is a duplicate
+                }
+                if (message.type() != WebhookMessage.WebhookMessageType.TEXT) {
+                  LOGGER.warn(
+                      "received message of type '"
+                          + message.type()
+                          + "', only able to handle text messages at this time");
+                  continue;
+                }
+                TextWebhookMessage textMessage = (TextWebhookMessage) message;
+                waMessages.add(
+                    new WAMessage(
+                        message.timestamp(),
+                        message.id(),
+                        message.from(),
+                        phoneNumberId,
+                        textMessage.text().body(),
+                        Message.Role.USER));
+                readExecutor.execute(() -> markRead(phoneNumberId, textMessage.id().toString()));
+              }
+            });
+
+    return waMessages;
+  }
+
+  @TestOnly
+  @This
+  WAMessageHandler baseUrlFactory(Function<Identifier, URI> baseURLFactory) {
+    this.baseURLFactory = baseURLFactory;
+    return this;
+  }
+
+  @Override
+  public void respond(WAMessage message) throws IOException {
+    for (String text : CHUNKER.chunks(message.message()).toList()) {
+      send(message.recipientId(), message.senderId(), text);
+    }
+  }
+
+  private void send(Identifier recipient, Identifier sender, String text) throws IOException {
+    ObjectNode body =
+        MAPPER
+            .createObjectNode()
+            .put("recipient_type", "individual")
```

Comment:
Probably best practice to have it but I believe recipient_type is an optional field that can only be one value (https://developers.facebook.com/docs/whatsapp/cloud-api/reference/messages#message-object)

Reply:
correct, I'm just defending against the possibility of us introducing group messages